File: packages/save_file/lib/src/dto/save_file_dto.dart
import 'package:utopia_save_file/src/model/save_file_metadata.dart';

sealed class SaveFileDto {
  final String name, mime;

  SaveFileDto(SaveFileMetadata metadata)
      : name = metadata.name,
        mime = metadata.mime;

  Map<String, dynamic> toMap() => {"name": name, "mime": mime};

  static const fromFile = SaveFileDtoFromFile.new;
  static const fromUrl = SaveFileDtoFromUrl.new;
  static const fromBytes = SaveFileDtoFromBytes.new;
}

final class SaveFileDtoFromFile extends SaveFileDto {
  final String path;

  SaveFileDtoFromFile(super.metadata, {required this.path});

  @override
  Map<String, dynamic> toMap() => {...super.toMap(), "path": path};
}

final class SaveFileDtoFromUrl extends SaveFileDto {
  final String url;

  SaveFileDtoFromUrl(super.metadata, {required this.url});

  @override
  Map<String, dynamic> toMap() => {...super.toMap(), "url": url};
}

final class SaveFileDtoFromBytes extends SaveFileDto {
  final List<int> bytes;

  SaveFileDtoFromBytes(super.metadata, {required this.bytes});

  @override
  Map<String, dynamic> toMap() => {...super.toMap(), "bytes": bytes};
}

File: packages/save_file/android/src/main/kotlin/com/utopiaultimate/flutter/utopia_save_file/dto/SaveFileDto.kt
package com.utopiaultimate.flutter.utopia_save_file.dto

sealed class SaveFileDto(map: Map<String, Any>) {
  val name: String by map
  val mime: String by map

  class FromFile(map: Map<String, Any>) : SaveFileDto(map) {
    val path: String by map
  }

  class FromUrl(map: Map<String, Any>) : SaveFileDto(map) {
    val url: String by map
  }

  class FromBytes(map: Map<String, Any>) : SaveFileDto(map) {
    val bytes: ByteArray by map
  }
}